src/js_udp_wrap.cc

```diff
@@ -0,0 +1,223 @@
+    buffers[i] = Buffer::Copy(env(), bufs[i].base, bufs[i].len)
+        .ToLocalChecked();
+    total_len += bufs[i].len;
+  }
+
+  Local<Value> args[] = {
+    listener()->CreateSendWrap(total_len)->object(),
+    Array::New(env()->isolate(), buffers.out(), nbufs),
+    AddressToJS(env(), addr)
+  };
+
+  if (!MakeCallback(env()->onwrite_string(), arraysize(args), args)
+          .ToLocal(&value) ||
+      !value->IntegerValue(env()->context()).To(&value_int)) {
+    if (try_catch.HasCaught() && !try_catch.HasTerminated())
+      errors::TriggerUncaughtException(env()->isolate(), try_catch);
+  }
+  return value_int;
+}
+
+int JSUDPWrap::GetPeerName(sockaddr* name, int* namelen) {
+  return GetSockaddr(name, namelen, true);
+}
+
+int JSUDPWrap::GetSockName(sockaddr* name, int* namelen) {
+  return GetSockaddr(name, namelen, false);
+}
+
+int JSUDPWrap::GetSockaddr(sockaddr* name, int* namelen, bool peer) {
+  // TODO(addaleax): Maybe turn this into a real JS-based method.
+  sockaddr_in addr_in;
+  CHECK_EQ(uv_ip4_addr("127.0.0.1", 1337, &addr_in), 0);
+  memcpy(name, &addr_in,
+         std::min(static_cast<size_t>(*namelen), sizeof(addr_in)));
+  *namelen = sizeof(addr_in);
+  return 0;
+}
+
+void JSUDPWrap::New(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  CHECK(args.IsConstructCall());
+  new JSUDPWrap(env, args.Holder());
+}
+
+void JSUDPWrap::EmitReceived(const FunctionCallbackInfo<Value>& args) {
+  JSUDPWrap* wrap;
+  ASSIGN_OR_RETURN_UNWRAP(&wrap, args.Holder());
+  Environment* env = wrap->env();
+
+  ArrayBufferViewContents<char> buffer(args[0]);
+  const char* data = buffer.data();
+  int len = buffer.length();
+
+  CHECK(args[1]->IsInt32());   // family
+  CHECK(args[2]->IsString());  // address
+  CHECK(args[3]->IsInt32());   // port
+  CHECK(args[4]->IsInt32());   // flags
+  int family = args[1].As<Int32>()->Value() == 4 ? AF_INET : AF_INET6;
+  Utf8Value address(env->isolate(), args[2]);
+  int port = args[3].As<Int32>()->Value();
+  int flags = args[3].As<Int32>()->Value();
+
+  sockaddr_storage addr;
+  CHECK_EQ(sockaddr_for_family(family, *address, port, &addr), 0);
+
+  // Repeatedly ask the stream's owner for memory, copy the data that we
+  // just read from JS into those buffers and emit them as reads.
+  while (len != 0) {
+    uv_buf_t buf = wrap->listener()->OnAlloc(len);
+    ssize_t avail = len;
+    if (static_cast<ssize_t>(buf.len) < avail)
+      avail = buf.len;
+
+    memcpy(buf.base, data, avail);
+    data += avail;
+    len -= avail;
+    wrap->listener()->OnRecv(
+        avail, buf, reinterpret_cast<sockaddr*>(&addr), flags);
+  }
+}
+
+void JSUDPWrap::OnSendDone(const FunctionCallbackInfo<Value>& args) {
+  JSUDPWrap* wrap;
+  ASSIGN_OR_RETURN_UNWRAP(&wrap, args.Holder());
+
+  CHECK(args[0]->IsObject());
+  CHECK(args[1]->IsInt32());
+  ReqWrap<uv_udp_send_t>* req_wrap;
+  ASSIGN_OR_RETURN_UNWRAP(&wrap, args[0].As<Object>());
```

Comment:
Should this be `&req_wrap` instead of `&wrap`?

Reply:
```suggestion
  ASSIGN_OR_RETURN_UNWRAP(&req_wrap, args[0].As<Object>());
```